File: code/python/analysis_FeaturePredictionAccuracy.py
'''
Feature prediction accuracy (image_seen, category_seen, category_imagined)

This file is a part of GenericDecoding_demo.
'''


import os
import pickle
import sys
from itertools import product
from time import time

import numpy as np
import pandas as pd

sys.path.append('lib/bdpy') # Path to bdpy
import bdpy
from bdpy.stats import corrcoef

import gd_parameters as gd
from gd_features import Features


#----------------------------------------------------------------------#
# Global settings                                                      #
#----------------------------------------------------------------------#

analysis_name = __file__

# Get global parametes from gd_parameters
data_dir = gd.data_dir
subject_list = gd.subject_list
roi_list = gd.roi_list
feature_file = gd.feature_file
feature_type = gd.feature_type
result_dir = gd.result_dir
feature_file = gd.feature_file
feature_type = gd.feature_type
result_dir = gd.result_dir

result_file = gd.results_featurepred


#----------------------------------------------------------------------#
# Main                                                                 #
#----------------------------------------------------------------------#

if __name__ == '__main__':
    print 'Running ' + analysis_name

    ## Load data (image features) --------------------------------------
    print 'Loading image features'
    features = Features(os.path.join(data_dir, feature_file), feature_type)

    ## Create result dir -----------------------------------------------
    if not os.path.isdir(result_dir):
        os.makedirs(result_dir)

    ## Run analysis for each subject, ROI, and feature -----------------
    results_columns = ['subject', 'roi', 'feature',
                       'category_test_percept', 'category_test_imagery',
                       'predict_percept', 'predict_imagery',
                       'predacc_image_percept',
                       'predacc_category_percept',
                       'predacc_category_imagery']

    results = pd.DataFrame([], columns=results_columns)

    for sbj, roi, feat in product(subject_list, roi_list, features.layers):
        start_time = time()

        analysis_id = '%s-%s-%s-%s' % (__file__, sbj, roi, feat)
        print 'Analysis %s' % analysis_id

        ## Load feature prediction results for each unit
        result_unit_file = os.path.join(result_dir, sbj, roi, feat + '.pkl')
        with open(result_unit_file, 'rb') as f:
            results_unit = pickle.load(f)

        ## Preparing data (image features)
        feature = features.get('value', feat)
        feature_type = features.get('type')
        feature_catlabel = features.get('category_label')

        ## Calculate image and category feature prediction accuracy

        ## Aggregate all units prediction (num_sample * num_unit)
        pred_percept = np.vstack(results_unit['predict_percept_catave']).T
        pred_imagery = np.vstack(results_unit['predict_imagery_catave']).T

        cat_percept = results_unit['category_test_percept'][0]
        cat_imagery = results_unit['category_test_imagery'][0]

        ## Get image features for image_seen, category_seen, and category_imagined
        ind_imgtest = feature_type == 2
        ind_cattest = feature_type == 3

        test_feat_img = bdpy.get_refdata(feature[ind_imgtest, :],
                                         feature_catlabel[ind_imgtest],
                                         cat_percept);
        test_feat_cat_percept = bdpy.get_refdata(feature[ind_cattest, :],
                                                 feature_catlabel[ind_cattest],
                                                 cat_percept)
        test_feat_cat_imagery = bdpy.get_refdata(feature[ind_cattest, :],
                                                 feature_catlabel[ind_cattest],
                                                 cat_imagery)

        ## Get image and category feature prediction accuracy
        predacc_image_percept = np.nanmean(corrcoef(pred_percept, test_feat_img, var='col'))
        predacc_category_percept = np.nanmean(corrcoef(pred_percept, test_feat_cat_percept, var='col'))
        predacc_category_imagery = np.nanmean(corrcoef(pred_imagery, test_feat_cat_imagery, var='col'))

        print 'Pred acc (image_percpet):    %f' % predacc_image_percept
        print 'Pred acc (category_percpet): %f' % predacc_category_percept
        print 'Pred acc (category_imagery): %f' % predacc_category_imagery
        
        df_tmp = pd.DataFrame({'subject' : [sbj],
                               'roi' : [roi],
                               'feature' : [feat],
                               'category_test_percept' : [cat_percept],
                               'category_test_imagery' : [cat_imagery],
                               'predict_percept' : [pred_percept],
                               'predict_imagery' : [pred_imagery],
                               'predacc_image_percept' : [predacc_image_percept],
                               'predacc_category_percept' : [predacc_category_percept],
                               'predacc_category_imagery' : [predacc_category_imagery]},
                              columns=results_columns)
        results = results.append(df_tmp, ignore_index=True)

        print 'Time: %.3f sec' % (time() - start_time)

    ## Save results ----------------------------------------------------
    print 'Saving %s' % os.path.join(result_dir, result_file)
    with open(os.path.join(result_dir, result_file), 'wb') as f:
        pickle.dump(results, f)

File: code/matlab/analysis_FeaturePredictionAccuracy.m
% analysis_FeaturePredictionAccuracy    Calculate feature prediction accuracy
%
% Author: Tomoyasu Horikawa <horikawa-t@atr.jp>, Shuntaro C. Aoki <aoki@atr.jp>
%


clear all;


%% Initial settings %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Data settings
% subjectList  : List of subject IDs [cell array]
% featureList  : List of image features [cell array]
% roiList      : List of ROIs [cell array]

subjectList = {'Subject1', 'Subject2', 'Subject3', 'Subject4', 'Subject5'};
featureList = {'cnn1', 'cnn2', 'cnn3', 'cnn4', ...
               'cnn5', 'cnn6', 'cnn7', 'cnn8', ...
               'hmax1', 'hmax2', 'hmax3', 'gist', 'sift'};
roiList     = {'V1', 'V2', 'V3', 'V4', 'FFA', 'LOC', 'PPA', 'LVC', 'HVC',  'VC'};

% Image feature data
imageFeatureFile = 'ImageFeatures.mat';

%% Directory settings
workDir = pwd;
dataDir = fullfile(workDir, 'data');       % Directory containing brain and image feature data
resultsDir = fullfile(workDir, 'results'); % Directory to save analysis results

%% File name settings
predResultFileNameFormat = @(s, r, f) fullfile(resultsDir, sprintf('%s/%s/%s.mat', s, r, f));
resultFile = fullfile(resultsDir, 'FeaturePrediction.mat');


%% Analysis Main %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

fprintf('%s started\n', mfilename);

%%----------------------------------------------------------------------
%% Initialization
%%----------------------------------------------------------------------

addpath(genpath('./lib'));

if ~exist(resultsDir, 'dir'), mkdir(resultsDir); end;

%%----------------------------------------------------------------------
%% Load data
%%----------------------------------------------------------------------

%% Load image features
fprintf('Loading image feature data...\n');

[feat.dataSet, feat.metaData] = load_data(fullfile(dataDir, imageFeatureFile));

%%----------------------------------------------------------------------
%% Create analysis parameter matrix (analysisParam)
%%----------------------------------------------------------------------

analysisParam = uint16(zeros(length(subjectList) * length(roiList) * length(featureList), 3));

c = 1;
for iSbj = 1:length(subjectList)
for iRoi = 1:length(roiList)
for iFeat = 1:length(featureList)
    analysisParam(c, :) = [ iSbj, iRoi, iFeat ];
    c =  c + 1;
end
end
end

%%----------------------------------------------------------------------
%% Analysis loop
%%----------------------------------------------------------------------

results = [];

for n = 1:size(analysisParam, 1)

    %% Initialization --------------------------------------------------

    % Get data index in the current analysis
    iSbj = analysisParam(n, 1);
    iRoi = analysisParam(n, 2);
    iFeat = analysisParam(n, 3);

    % Set analysis ID and a result file name
    analysisId = sprintf('%s-%s-%s-%s', ...
                         mfilename, ...
                         subjectList{iSbj}, ...
                         roiList{iRoi}, ...
                         featureList{iFeat});

    % Check or double-running
    if checkfiles(resultFile)
        fprintf('The analysis is already done and skipped\n');
        continue;
    end

    fprintf('Start %s\n', analysisId);

    %% Get image features ----------------------------------------------
    layerFeat = select_feature(feat.dataSet, feat.metaData, sprintf('%s = 1', featureList{iFeat}));
    catIds = get_dataset(feat.dataSet, feat.metaData, 'CatID');
    featType = get_dataset(feat.dataSet, feat.metaData, 'FeatureType');

    %% Aggregate feature units -----------------------------------------
    predResultFile = predResultFileNameFormat(subjectList{iSbj}, roiList{iRoi}, featureList{iFeat});
    res = load(predResultFile);

    predPercept = res.predictPerceptAveraged;
    predImagery = res.predictImageryAveraged;

    categoryPercept = res.categoryTestPercept;
    categoryImagery = res.categoryTestImagery;

    %% Calculate category feature prediction accuracy ------------------

    %% Get test features (images)
    testFeat = layerFeat(featType == 2, :);
    testCatIds = catIds(featType == 2, :);

    testFeat = get_refdata(testFeat, testCatIds, categoryPercept);

    %% Image feature prediction accuracy (profile correlation)
    predAcc.image.perception = nanmean(diag(fastcorr(predPercept, testFeat)));

    %% Get test features (category averaged)
    catTestFeat = layerFeat(featType == 3, :);
    catTestCatIds = catIds(featType == 3, :);

    catTestFeatPercept = get_refdata(catTestFeat, catTestCatIds, categoryPercept);
    catTestFeatImagery = get_refdata(catTestFeat, catTestCatIds, categoryImagery);

    %% Category-average feature prediction accuracy (profile correlation)
    predAcc.category.perception = nanmean(diag(fastcorr(predPercept, catTestFeatPercept)));
    predAcc.category.imagery    = nanmean(diag(fastcorr(predImagery, catTestFeatImagery)));

    results(n).subject = subjectList{iSbj};
    results(n).roi = roiList{iRoi};
    results(n).feature = featureList{iFeat};
    results(n).categoryTestPercept = categoryPercept;
    results(n).categoryTestImagery = categoryImagery;
    results(n).predictPercept = predPercept;
    results(n).predictImagery = predImagery;
    results(n).predaccImagePercept = predAcc.image.perception;
    results(n).predaccCategoryPercept = predAcc.category.perception;
    results(n).predaccCategoryImagery = predAcc.category.imagery;
end

%% Save data -----------------------------------------------------------
save(resultFile, 'results', '-v7.3');

fprintf('%s done\n', mfilename);
